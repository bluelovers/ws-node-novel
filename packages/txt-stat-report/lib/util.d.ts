/**
 * Created by user on 2019/2/23.
 */
export declare const regexpPunctuation: RegExp;
export declare function removePunctuation(input: string): string;
export declare function removeSpace(input: string): string;
export declare function removeLine(input: string): string;
/**
 * 合併多個 regexp 為一個
 */
export declare function regexMerge<T extends RegExp>(list: T[]): RegExp;
export declare function removeBom(input: string): string;
