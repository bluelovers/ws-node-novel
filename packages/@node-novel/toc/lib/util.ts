/**
 * Created by user on 2018/11/14/014.
 */

import { _trim, createSortCallback, defaultSortCallback, EnumToLowerCase } from '@node-novel/sort';
import { array_unique } from 'array-hyper-unique';
import FastGlob = require('fast-glob');
import fs = require('fs-iconv');
import { IMdconfMeta, mdconf_parse } from 'node-novel-info';
import BluebirdPromise = require('bluebird');
import StrUtil = require('str-util');

export async function loadReadmeMeta<T extends IMdconfMeta = IMdconfMeta>(file: string): Promise<T>
{
	return fs.readFile(file)
		.then(function (data)
		{
			return mdconf_parse(data, {
				// 當沒有包含必要的內容時不產生錯誤
				throw: false,
				// 允許不標準的 info 內容
				lowCheckLevel: true,
			});
		})
		.catch(function ()
		{
			return null;
		})
		;
}

export function loadReadmeMetaSync<T extends IMdconfMeta = IMdconfMeta>(file: string): T
{
	try
	{
		let data = fs.readFileSync(file);

		// @ts-ignore
		return mdconf_parse(data, {
			// 當沒有包含必要的內容時不產生錯誤
			throw: false,
			// 允許不標準的 info 內容
			lowCheckLevel: true,
		})
	}
	catch (e)
	{

	}

	return null;
}

export function getNovelTitles<T extends IMdconfMeta = IMdconfMeta>(meta: T): string[]
{
	if (meta && meta.novel)
	{
		let arr = [
				'title',
				'title_source',
				'title_jp',
				'title_ja',
				'title_zh',
				'title_tw',
				'title_cn',
			].concat(Object.keys(meta.novel))
			.reduce(function (a, key: string)
			{
				if (key.indexOf('title') === 0)
				{
					a.push(meta.novel[key])
				}

				return a
			}, [])
		;

		if (meta.novel.series)
		{
			arr.push(meta.novel.series.name);
			arr.push(meta.novel.series.name_short);
		}

		arr = array_unique(arr.filter(v => v && ![
			'undefined',
			'長編 【連載】',
			'連載中',
		].includes(v)));

		return arr;
	}

	return [];
}

export function globFirst(...argv: Parameters<typeof FastGlob["stream"]>): BluebirdPromise<string>
{
	return new BluebirdPromise<string>(function (resolve, reject)
	{
		let fgs = FastGlob.stream(...argv);

		fgs.on('data', (entry) =>
		{
			resolve(entry);

			// @ts-ignore
			fgs.destroy();
		});
		fgs.once('error', reject);
		fgs.once('end', () => resolve(undefined));
	})
}

export function md_href(href: string)
{
	return href.split('/').map(encodeURIComponent).join('/');
}

export function md_anchor_gitee(title: string)
{
	let anchor = title
		.replace(/[a-z]+/ig, function (s)
		{
			return s.toLowerCase();
		})
		.replace(/[\.．\/／　＠@（）\(\)～~]/g, '')
		.replace(/[ ]/g, '-')
	;

	return md_href(anchor);
}

export function md_link_escape(text: string)
{
	return text.replace(/[\[\]]/g, function (s)
	{
		return '\\' + s;
	})
}

export const tocSortCallback = createSortCallback({
	dotNum: true,
	transpileBase(input: string, isSub?: any)
	{
		let s = StrUtil.toHalfWidth(input);
		return s
	},
	toLowerCase: EnumToLowerCase.toLocaleLowerCase,
});
