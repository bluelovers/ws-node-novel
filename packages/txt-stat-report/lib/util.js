"use strict";
/**
 * Created by user on 2019/2/23.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const array_hyper_unique_1 = require("array-hyper-unique");
exports.regexpPunctuation = regexMerge([
    /\p{Punctuation}+/gu,
    /[\u2000-\u206F\u2E00-\u2E7F\uff00-\uff0f\uff1a-\uff20\uff3b-\uff40\uff5b-\uff65\uffe0-\uffef\u2500-\u257f\u2200-\u22ff\u25A0-\u25FF\u2600-\u26F0\u2190-\u21FF\u02b9-\u02df\u02E4-\u02f0\u2580-\u259F]+/ug,
    /[\u0021-\u002F\u003A-\u0040\u005B-\u0060\u007B-\u007F]/gu,
    /[\u00A1-\u00BF\u00D7\u00F7]/gu,
    /[\u2100-\u214F]/gu,
]);
function removePunctuation(input) {
    return input
        .replace(exports.regexpPunctuation, '');
}
exports.removePunctuation = removePunctuation;
function removeSpace(input) {
    return input
        .replace(/\s+/g, function (s) {
        return s.replace(/[^\r\n]+/g, '');
    })
        .replace(/[\xA0 　]+/gu, '');
}
exports.removeSpace = removeSpace;
function removeLine(input) {
    return input
        .replace(/[\r\n]+/gu, '');
}
exports.removeLine = removeLine;
/**
 * 合併多個 regexp 為一個
 */
function regexMerge(list) {
    let source = [];
    let flags = [];
    list.forEach(function (a) {
        source.push(a.source);
        a.flags && flags.push(...a.flags.split(''));
    });
    array_hyper_unique_1.array_unique_overwrite(source);
    array_hyper_unique_1.array_unique_overwrite(flags);
    return new RegExp(source.join('|'), flags.join(''));
}
exports.regexMerge = regexMerge;
function removeBom(input) {
    return input
        .replace(/\uFEFF/gu, '');
}
exports.removeBom = removeBom;
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoidXRpbC5qcyIsInNvdXJjZVJvb3QiOiIiLCJzb3VyY2VzIjpbInV0aWwudHMiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IjtBQUFBOztHQUVHOztBQUVILDJEQUEwRTtBQUU3RCxRQUFBLGlCQUFpQixHQUFHLFVBQVUsQ0FBQztJQUMzQyxvQkFBb0I7SUFDcEIsME1BQTBNO0lBQzFNLDBEQUEwRDtJQUMxRCwrQkFBK0I7SUFDL0IsbUJBQW1CO0NBQ25CLENBQUMsQ0FBQztBQUVILFNBQWdCLGlCQUFpQixDQUFDLEtBQWE7SUFFOUMsT0FBTyxLQUFLO1NBQ1YsT0FBTyxDQUFDLHlCQUFpQixFQUFFLEVBQUUsQ0FBQyxDQUs5QjtBQUNILENBQUM7QUFURCw4Q0FTQztBQUVELFNBQWdCLFdBQVcsQ0FBQyxLQUFhO0lBRXhDLE9BQU8sS0FBSztTQUNWLE9BQU8sQ0FBQyxNQUFNLEVBQUUsVUFBVSxDQUFDO1FBRTNCLE9BQU8sQ0FBQyxDQUFDLE9BQU8sQ0FBQyxXQUFXLEVBQUUsRUFBRSxDQUFDLENBQUE7SUFDbEMsQ0FBQyxDQUFDO1NBQ0QsT0FBTyxDQUFDLGFBQWEsRUFBRSxFQUFFLENBQUMsQ0FDMUI7QUFDSCxDQUFDO0FBVEQsa0NBU0M7QUFFRCxTQUFnQixVQUFVLENBQUMsS0FBYTtJQUV2QyxPQUFPLEtBQUs7U0FDVixPQUFPLENBQUMsV0FBVyxFQUFFLEVBQUUsQ0FBQyxDQUN4QjtBQUNILENBQUM7QUFMRCxnQ0FLQztBQUVEOztHQUVHO0FBQ0gsU0FBZ0IsVUFBVSxDQUFtQixJQUFTO0lBRXJELElBQUksTUFBTSxHQUFhLEVBQUUsQ0FBQztJQUMxQixJQUFJLEtBQUssR0FBYSxFQUFFLENBQUM7SUFFekIsSUFBSSxDQUFDLE9BQU8sQ0FBQyxVQUFVLENBQUM7UUFFdkIsTUFBTSxDQUFDLElBQUksQ0FBQyxDQUFDLENBQUMsTUFBTSxDQUFDLENBQUM7UUFFdEIsQ0FBQyxDQUFDLEtBQUssSUFBSSxLQUFLLENBQUMsSUFBSSxDQUFDLEdBQUcsQ0FBQyxDQUFDLEtBQUssQ0FBQyxLQUFLLENBQUMsRUFBRSxDQUFDLENBQUMsQ0FBQztJQUM3QyxDQUFDLENBQUMsQ0FBQztJQUVILDJDQUFzQixDQUFDLE1BQU0sQ0FBQyxDQUFDO0lBQy9CLDJDQUFzQixDQUFDLEtBQUssQ0FBQyxDQUFDO0lBRTlCLE9BQU8sSUFBSSxNQUFNLENBQUMsTUFBTSxDQUFDLElBQUksQ0FBQyxHQUFHLENBQUMsRUFBRSxLQUFLLENBQUMsSUFBSSxDQUFDLEVBQUUsQ0FBQyxDQUFDLENBQUM7QUFDckQsQ0FBQztBQWhCRCxnQ0FnQkM7QUFFRCxTQUFnQixTQUFTLENBQUMsS0FBYTtJQUV0QyxPQUFPLEtBQUs7U0FDVixPQUFPLENBQUMsVUFBVSxFQUFFLEVBQUUsQ0FBQyxDQUN2QjtBQUNILENBQUM7QUFMRCw4QkFLQyIsInNvdXJjZXNDb250ZW50IjpbIi8qKlxuICogQ3JlYXRlZCBieSB1c2VyIG9uIDIwMTkvMi8yMy5cbiAqL1xuXG5pbXBvcnQgeyBhcnJheV91bmlxdWUsIGFycmF5X3VuaXF1ZV9vdmVyd3JpdGUgfSBmcm9tICdhcnJheS1oeXBlci11bmlxdWUnO1xuXG5leHBvcnQgY29uc3QgcmVnZXhwUHVuY3R1YXRpb24gPSByZWdleE1lcmdlKFtcblx0L1xccHtQdW5jdHVhdGlvbn0rL2d1LFxuXHQvW1xcdTIwMDAtXFx1MjA2RlxcdTJFMDAtXFx1MkU3RlxcdWZmMDAtXFx1ZmYwZlxcdWZmMWEtXFx1ZmYyMFxcdWZmM2ItXFx1ZmY0MFxcdWZmNWItXFx1ZmY2NVxcdWZmZTAtXFx1ZmZlZlxcdTI1MDAtXFx1MjU3ZlxcdTIyMDAtXFx1MjJmZlxcdTI1QTAtXFx1MjVGRlxcdTI2MDAtXFx1MjZGMFxcdTIxOTAtXFx1MjFGRlxcdTAyYjktXFx1MDJkZlxcdTAyRTQtXFx1MDJmMFxcdTI1ODAtXFx1MjU5Rl0rL3VnLFxuXHQvW1xcdTAwMjEtXFx1MDAyRlxcdTAwM0EtXFx1MDA0MFxcdTAwNUItXFx1MDA2MFxcdTAwN0ItXFx1MDA3Rl0vZ3UsXG5cdC9bXFx1MDBBMS1cXHUwMEJGXFx1MDBEN1xcdTAwRjddL2d1LFxuXHQvW1xcdTIxMDAtXFx1MjE0Rl0vZ3UsXG5dKTtcblxuZXhwb3J0IGZ1bmN0aW9uIHJlbW92ZVB1bmN0dWF0aW9uKGlucHV0OiBzdHJpbmcpXG57XG5cdHJldHVybiBpbnB1dFxuXHRcdC5yZXBsYWNlKHJlZ2V4cFB1bmN0dWF0aW9uLCAnJylcblx0XHQvLy5yZXBsYWNlKC9cXHB7UHVuY3R1YXRpb259Ky9ndSwgJycpXG5cdFx0Ly8ucmVwbGFjZSgvW1xcdTIwMDAtXFx1MjA2RlxcdTJFMDAtXFx1MkU3RlxcdWZmMDAtXFx1ZmYwZlxcdWZmMWEtXFx1ZmYyMFxcdWZmM2ItXFx1ZmY0MFxcdWZmNWItXFx1ZmY2NVxcdWZmZTAtXFx1ZmZlZlxcdTI1MDAtXFx1MjU3ZlxcdTIyMDAtXFx1MjJmZlxcdTI1QTAtXFx1MjVGRlxcdTI2MDAtXFx1MjZGMFxcdTIxOTAtXFx1MjFGRlxcdTAyYjktXFx1MDJkZlxcdTAyRTQtXFx1MDJmMFxcdTI1ODAtXFx1MjU5Rl0vdWcsICcnKVxuXHRcdC8vLnJlcGxhY2UoL1vii69dKy9ndSwgJycpXG5cdFx0Ly8ucmVwbGFjZSgvWychXCIjJCUmKCkqKyxcXC0uXFwvXFxcXDo7PD0+P0BcXFtcXF1eX2B7fH1+wrDihJbDt8OX4oiSwqZdL3VnLCAnJylcblx0XHQ7XG59XG5cbmV4cG9ydCBmdW5jdGlvbiByZW1vdmVTcGFjZShpbnB1dDogc3RyaW5nKVxue1xuXHRyZXR1cm4gaW5wdXRcblx0XHQucmVwbGFjZSgvXFxzKy9nLCBmdW5jdGlvbiAocylcblx0XHR7XG5cdFx0XHRyZXR1cm4gcy5yZXBsYWNlKC9bXlxcclxcbl0rL2csICcnKVxuXHRcdH0pXG5cdFx0LnJlcGxhY2UoL1tcXHhBMCDjgIBdKy9ndSwgJycpXG5cdFx0O1xufVxuXG5leHBvcnQgZnVuY3Rpb24gcmVtb3ZlTGluZShpbnB1dDogc3RyaW5nKVxue1xuXHRyZXR1cm4gaW5wdXRcblx0XHQucmVwbGFjZSgvW1xcclxcbl0rL2d1LCAnJylcblx0XHQ7XG59XG5cbi8qKlxuICog5ZCI5L215aSa5YCLIHJlZ2V4cCDngrrkuIDlgItcbiAqL1xuZXhwb3J0IGZ1bmN0aW9uIHJlZ2V4TWVyZ2U8VCBleHRlbmRzIFJlZ0V4cD4obGlzdDogVFtdKVxue1xuXHRsZXQgc291cmNlOiBzdHJpbmdbXSA9IFtdO1xuXHRsZXQgZmxhZ3M6IHN0cmluZ1tdID0gW107XG5cblx0bGlzdC5mb3JFYWNoKGZ1bmN0aW9uIChhKVxuXHR7XG5cdFx0c291cmNlLnB1c2goYS5zb3VyY2UpO1xuXG5cdFx0YS5mbGFncyAmJiBmbGFncy5wdXNoKC4uLmEuZmxhZ3Muc3BsaXQoJycpKTtcblx0fSk7XG5cblx0YXJyYXlfdW5pcXVlX292ZXJ3cml0ZShzb3VyY2UpO1xuXHRhcnJheV91bmlxdWVfb3ZlcndyaXRlKGZsYWdzKTtcblxuXHRyZXR1cm4gbmV3IFJlZ0V4cChzb3VyY2Uuam9pbignfCcpLCBmbGFncy5qb2luKCcnKSk7XG59XG5cbmV4cG9ydCBmdW5jdGlvbiByZW1vdmVCb20oaW5wdXQ6IHN0cmluZylcbntcblx0cmV0dXJuIGlucHV0XG5cdFx0LnJlcGxhY2UoL1xcdUZFRkYvZ3UsICcnKVxuXHRcdDtcbn1cbiJdfQ==