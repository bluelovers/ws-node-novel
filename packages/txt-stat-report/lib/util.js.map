{"version":3,"file":"util.js","sourceRoot":"","sources":["util.ts"],"names":[],"mappings":";AAAA;;GAEG;;AAEH,2DAA0E;AAE7D,QAAA,iBAAiB,GAAG,UAAU,CAAC;IAC3C,oBAAoB;IACpB,0MAA0M;IAC1M,0DAA0D;IAC1D,+BAA+B;IAC/B,mBAAmB;CACnB,CAAC,CAAC;AAEH,SAAgB,iBAAiB,CAAC,KAAa;IAE9C,OAAO,KAAK;SACV,OAAO,CAAC,yBAAiB,EAAE,EAAE,CAAC,CAK9B;AACH,CAAC;AATD,8CASC;AAED,SAAgB,WAAW,CAAC,KAAa;IAExC,OAAO,KAAK;SACV,OAAO,CAAC,MAAM,EAAE,UAAU,CAAC;QAE3B,OAAO,CAAC,CAAC,OAAO,CAAC,WAAW,EAAE,EAAE,CAAC,CAAA;IAClC,CAAC,CAAC;SACD,OAAO,CAAC,aAAa,EAAE,EAAE,CAAC,CAC1B;AACH,CAAC;AATD,kCASC;AAED,SAAgB,UAAU,CAAC,KAAa;IAEvC,OAAO,KAAK;SACV,OAAO,CAAC,WAAW,EAAE,EAAE,CAAC,CACxB;AACH,CAAC;AALD,gCAKC;AAED;;GAEG;AACH,SAAgB,UAAU,CAAmB,IAAS;IAErD,IAAI,MAAM,GAAa,EAAE,CAAC;IAC1B,IAAI,KAAK,GAAa,EAAE,CAAC;IAEzB,IAAI,CAAC,OAAO,CAAC,UAAU,CAAC;QAEvB,MAAM,CAAC,IAAI,CAAC,CAAC,CAAC,MAAM,CAAC,CAAC;QAEtB,CAAC,CAAC,KAAK,IAAI,KAAK,CAAC,IAAI,CAAC,GAAG,CAAC,CAAC,KAAK,CAAC,KAAK,CAAC,EAAE,CAAC,CAAC,CAAC;IAC7C,CAAC,CAAC,CAAC;IAEH,2CAAsB,CAAC,MAAM,CAAC,CAAC;IAC/B,2CAAsB,CAAC,KAAK,CAAC,CAAC;IAE9B,OAAO,IAAI,MAAM,CAAC,MAAM,CAAC,IAAI,CAAC,GAAG,CAAC,EAAE,KAAK,CAAC,IAAI,CAAC,EAAE,CAAC,CAAC,CAAC;AACrD,CAAC;AAhBD,gCAgBC;AAED,SAAgB,SAAS,CAAC,KAAa;IAEtC,OAAO,KAAK;SACV,OAAO,CAAC,UAAU,EAAE,EAAE,CAAC,CACvB;AACH,CAAC;AALD,8BAKC","sourcesContent":["/**\n * Created by user on 2019/2/23.\n */\n\nimport { array_unique, array_unique_overwrite } from 'array-hyper-unique';\n\nexport const regexpPunctuation = regexMerge([\n\t/\\p{Punctuation}+/gu,\n\t/[\\u2000-\\u206F\\u2E00-\\u2E7F\\uff00-\\uff0f\\uff1a-\\uff20\\uff3b-\\uff40\\uff5b-\\uff65\\uffe0-\\uffef\\u2500-\\u257f\\u2200-\\u22ff\\u25A0-\\u25FF\\u2600-\\u26F0\\u2190-\\u21FF\\u02b9-\\u02df\\u02E4-\\u02f0\\u2580-\\u259F]+/ug,\n\t/[\\u0021-\\u002F\\u003A-\\u0040\\u005B-\\u0060\\u007B-\\u007F]/gu,\n\t/[\\u00A1-\\u00BF\\u00D7\\u00F7]/gu,\n\t/[\\u2100-\\u214F]/gu,\n]);\n\nexport function removePunctuation(input: string)\n{\n\treturn input\n\t\t.replace(regexpPunctuation, '')\n\t\t//.replace(/\\p{Punctuation}+/gu, '')\n\t\t//.replace(/[\\u2000-\\u206F\\u2E00-\\u2E7F\\uff00-\\uff0f\\uff1a-\\uff20\\uff3b-\\uff40\\uff5b-\\uff65\\uffe0-\\uffef\\u2500-\\u257f\\u2200-\\u22ff\\u25A0-\\u25FF\\u2600-\\u26F0\\u2190-\\u21FF\\u02b9-\\u02df\\u02E4-\\u02f0\\u2580-\\u259F]/ug, '')\n\t\t//.replace(/[⋯]+/gu, '')\n\t\t//.replace(/['!\"#$%&()*+,\\-.\\/\\\\:;<=>?@\\[\\]^_`{|}~°№÷×−¦]/ug, '')\n\t\t;\n}\n\nexport function removeSpace(input: string)\n{\n\treturn input\n\t\t.replace(/\\s+/g, function (s)\n\t\t{\n\t\t\treturn s.replace(/[^\\r\\n]+/g, '')\n\t\t})\n\t\t.replace(/[\\xA0 　]+/gu, '')\n\t\t;\n}\n\nexport function removeLine(input: string)\n{\n\treturn input\n\t\t.replace(/[\\r\\n]+/gu, '')\n\t\t;\n}\n\n/**\n * 合併多個 regexp 為一個\n */\nexport function regexMerge<T extends RegExp>(list: T[])\n{\n\tlet source: string[] = [];\n\tlet flags: string[] = [];\n\n\tlist.forEach(function (a)\n\t{\n\t\tsource.push(a.source);\n\n\t\ta.flags && flags.push(...a.flags.split(''));\n\t});\n\n\tarray_unique_overwrite(source);\n\tarray_unique_overwrite(flags);\n\n\treturn new RegExp(source.join('|'), flags.join(''));\n}\n\nexport function removeBom(input: string)\n{\n\treturn input\n\t\t.replace(/\\uFEFF/gu, '')\n\t\t;\n}\n"]}