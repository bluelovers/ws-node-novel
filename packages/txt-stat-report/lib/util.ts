/**
 * Created by user on 2019/2/23.
 */

import { array_unique, array_unique_overwrite } from 'array-hyper-unique';

export const regexpPunctuation = regexMerge([
	/\p{Punctuation}+/gu,
	/[\u2000-\u206F\u2E00-\u2E7F\uff00-\uff0f\uff1a-\uff20\uff3b-\uff40\uff5b-\uff65\uffe0-\uffef\u2500-\u257f\u2200-\u22ff\u25A0-\u25FF\u2600-\u26F0\u2190-\u21FF\u02b9-\u02df\u02E4-\u02f0\u2580-\u259F]+/ug,
	/[\u0021-\u002F\u003A-\u0040\u005B-\u0060\u007B-\u007F]/gu,
	/[\u00A1-\u00BF\u00D7\u00F7]/gu,
	/[\u2100-\u214F]/gu,
]);

export function removePunctuation(input: string)
{
	return input
		.replace(regexpPunctuation, '')
		//.replace(/\p{Punctuation}+/gu, '')
		//.replace(/[\u2000-\u206F\u2E00-\u2E7F\uff00-\uff0f\uff1a-\uff20\uff3b-\uff40\uff5b-\uff65\uffe0-\uffef\u2500-\u257f\u2200-\u22ff\u25A0-\u25FF\u2600-\u26F0\u2190-\u21FF\u02b9-\u02df\u02E4-\u02f0\u2580-\u259F]/ug, '')
		//.replace(/[⋯]+/gu, '')
		//.replace(/['!"#$%&()*+,\-.\/\\:;<=>?@\[\]^_`{|}~°№÷×−¦]/ug, '')
		;
}

export function removeSpace(input: string)
{
	return input
		.replace(/\s+/g, function (s)
		{
			return s.replace(/[^\r\n]+/g, '')
		})
		.replace(/[\xA0 　]+/gu, '')
		;
}

export function removeLine(input: string)
{
	return input
		.replace(/[\r\n]+/gu, '')
		;
}

/**
 * 合併多個 regexp 為一個
 */
export function regexMerge<T extends RegExp>(list: T[])
{
	let source: string[] = [];
	let flags: string[] = [];

	list.forEach(function (a)
	{
		source.push(a.source);

		a.flags && flags.push(...a.flags.split(''));
	});

	array_unique_overwrite(source);
	array_unique_overwrite(flags);

	return new RegExp(source.join('|'), flags.join(''));
}

export function removeBom(input: string)
{
	return input
		.replace(/\uFEFF/gu, '')
		;
}
