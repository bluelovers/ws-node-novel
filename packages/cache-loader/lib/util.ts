/**
 * Created by user on 2019/1/6/006.
 */

import moment from 'moment';
import { defaultSortCallback, naturalCompare } from '@node-novel/sort';
import { toHalfWidth } from '@lazy-cjk/fullhalf';
import sortObject from 'sort-object-keys2';
import { _trim } from '@node-novel/sort/lib/util';
import { EnumToLowerCase } from '@node-novel/sort/lib/types';
import { createSortCallback } from '@node-novel/sort/lib/core';

let defaultOffset = 8;

export { naturalCompare }

export function createMoment(...argv)
{
	return moment(...argv).utcOffset(defaultOffset);
}

const todayMoment = createMoment().startOf('day');

export let todayMomentTimestamp = todayMoment.valueOf();
export let todayMomentOffset = todayMoment.utcOffset();

export default todayMomentTimestamp

export function getTodayMomentTimestamp()
{
	return todayMoment.valueOf();
}

export function refreshTodayMoment()
{
	let k = createMoment()
		//.add(7, 'days')
		.startOf('day')
	;

	if (k.valueOf() != todayMoment.valueOf())
	{
		todayMoment.set(k.toObject());

		return true;
	}
}

Object.defineProperties(exports, {
	todayMomentTimestamp: {
		get: getTodayMomentTimestamp,
	},
	todayMomentOffset: {
		get()
		{
			return todayMoment.utcOffset();
		},
	},
});

export const cacheSortCallback = createSortCallback({
	dotNum: true,
	transpileBase(input: string, isSub?: any)
	{
		let s = toHalfWidth(input);
		return s
	},
	toLowerCase: EnumToLowerCase.toLocaleLowerCase,
});

export function freezeProperty<T>(who: T, prop: keyof T, freeze?: boolean)
{
	if (freeze)
	{
		try
		{
			// @ts-ignore
			who[prop] = Object.freeze(who[prop]);
		}
		catch (e)
		{

		}
	}

	Object.defineProperty(who, prop, {
		configurable: false,
		writable: false,
	});

	return who;
}

export function baseSortObject<T>(data: T): T
{
	return sortObject(data, {
		useSource: true,
		keys: Object.keys(data).sort(),
	});
}

export function tryRequireFS(): typeof import('fs-extra')
{
	let fs: typeof import('fs-extra');

	try
	{
		fs = require('fs-extra');
		return fs;
	}
	catch (e)
	{

	}
}

export function parsePathMainBase(pathMain: string)
{
	let is_out: boolean = null;
	let pathMain_base: string = undefined;
	let pathMain_out: string = undefined;

	if (pathMain != null)
	{
		let _m = pathMain.match(/^(.+?)(_out)?$/);

		is_out = !!_m[2];
		pathMain_base = _m[1];
		pathMain_out = pathMain_base + '_out';
	}

	return {
		is_out,
		pathMain_base,
		pathMain_out,
	}
}
