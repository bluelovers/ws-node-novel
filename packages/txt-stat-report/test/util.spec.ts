
import { readFileSync } from 'fs';
import { removePunctuation } from '../lib/util';
import { join } from 'path';
import { txtReport } from '../index';

const source = readFileSync(join(__dirname, 'res/punctuation.txt'))
	.toString()
;

test(`removePunctuation`, () =>
{
	let actual = removePunctuation(source.replace(/\s+/g, ''));
	let expected = '';

	expect(actual).toStrictEqual(expected);
	//expect(actual).toBeInstanceOf(Date);
	expect(actual).toMatchSnapshot();

});

test(`txtReport`, () =>
{
	let actual = txtReport(source);

	expect(actual).toMatchSnapshot();

});
