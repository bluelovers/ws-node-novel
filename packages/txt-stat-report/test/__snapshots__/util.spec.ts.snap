// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removePunctuation 1`] = `""`;

exports[`txtReport 1`] = `
Object {
  "buf_length": 343,
  "hanzi_length": 0,
  "ja_length": 1,
  "js_length": 142,
  "line_length": 3,
  "no_blank_line_length": 2,
  "punctuation_length": 139,
  "space_length": 0,
  "uni_length": 142,
}
`;
